File: end.java
import java.awt.*;

public class end extends klevel
{
    String[] text =  {
        "Ende",
        "",
        "Ulla gibt ein Bild",
        "von den Schmugglern",
        "dem Kommisar",
        "",
        "Die Schmuggler sind",
        "im Gefängnis und Larry",
        "bekommt die Belohnung"
    };

    double ty=300, ts=1, fontmax=40, fontmin=10;
    Point star[] = new Point[100];

    int stage = 1;  // 1 - text, 2  - move down, 9 - stop
    int ct = 0, ct_to = 8;
    boolean flash = false;

    public end ()
    {
        plot = "Ende";
        inst = "If you have 1000 pts, you can go to a bonus level.";
    }

    public void loadData ()
    {
        for (int i=0;i<100;i++) star[i] = new Point((int)(Math.random()*400),(int)(Math.random()*300));

        if (((koffer)master).score >= 1000) enableNextLevel();

        speed = 100;
        start();
    }

    public void drawStatic (Graphics g)
    {
        g.setColor(Color.black);
        g.fillRect(0,0,400,300);
    }

//  public void drawText (Graphics g, String s, Point p, int ha, int va, Color c, boolean three_d)

    public void drawNormal (Graphics g)
    {
        for (int i=0;i<100;i++)
        {
            g.setColor((Math.random()<.5)?Color.gray:Color.white);
            g.drawLine(star[i].x,star[i].y,star[i].x,star[i].y);
        }

        if (stage==1)
        {
            ty-=ts;
            double cy=ty;
            for (int i=0;i<text.length;i++)
            {
                if (cy<=300)
                {
                    g.setFont(new Font("Arial",Font.BOLD,(int)(fontmin+(fontmax-fontmin)*(cy/300))));
                    drawText(g,text[i],new Point(200,(int)cy),CENTER,BOTTOM, Color.yellow, true);
                    cy += g.getFontMetrics().getHeight();
                }
            }
            if (cy<10)
            {
                stage++;
                g.setFont(new Font("Arial",Font.BOLD,27));
            }
        }

        if (stage==2)
        {
            for (int i=0;i<100;i++)
            {
                int x = star[i].x, y = star[i].y;
                if ((y-=3)<0) y+=300;
                star[i] = new Point(x, y);
            }

            if (++ct>ct_to) { ct=0; flash=!flash; }
            if (flash)
            {
                String mes = (((koffer)master).score<1000) ? "Press RESTART" : "Go to Bonus Level!";
                drawText(g,mes,new Point(200,100),CENTER,TOP, Color.red, true);
            }
        }

    }



}

File: js/end.js
"use strict";

// Port of end.java.

class EndLevel extends KLevel {
  constructor(game) {
    super(game);
    this.plot = "Ende";
    this.inst = "If you have 1000 pts, you can go to a bonus level.";

    this.text = [
      "Ende",
      "",
      "Ulla gibt ein Bild",
      "von den Schmugglern",
      "dem Kommisar",
      "",
      "Die Schmuggler sind",
      "im Gefängnis und Larry",
      "bekommt die Belohnung",
    ];

    this.ty = 300;
    this.ts = 1;
    this.fontmax = 40;
    this.fontmin = 10;
    this.star = [];
    this.stage = 1;  // 1 - text, 2 - move down, 9 - stop
    this.ct = 0;
    this.ct_to = 8;
    this.flash = false;
  }

  async loadData() {
    for (let i = 0; i < 100; i++) {
      this.star[i] = {
        x: Math.floor(Math.random() * 400),
        y: Math.floor(Math.random() * 300),
      };
    }

    if (this.game.score >= 1000) this.enableNextLevel();

    this.speed = 100;
    this.start();
  }

  drawStatic(g) {
    g.fillStyle = "#000";
    g.fillRect(0, 0, 400, 300);
  }

  drawNormal(g) {
    for (let i = 0; i < 100; i++) {
      g.fillStyle = (Math.random() < 0.5) ? "gray" : "white";
      g.fillRect(this.star[i].x, this.star[i].y, 1, 1);
    }

    if (this.stage === 1) {
      this.ty -= this.ts;
      let cy = this.ty;
      for (const line of this.text) {
        if (cy <= 300) {
          const size = Math.floor(this.fontmin + (this.fontmax - this.fontmin) * (cy / 300));
          g.font = `bold ${Math.max(size, 1)}px Arial, sans-serif`;
          this.drawText(g, line, { x: 200, y: Math.trunc(cy) },
                        KLevel.CENTER, KLevel.BOTTOM, [255, 255, 0], true);
          const m = g.measureText("M");
          cy += (m.fontBoundingBoxAscent + m.fontBoundingBoxDescent) || size * 1.2;
        }
      }
      if (cy < 10) {
        this.stage++;
      }
    }

    if (this.stage === 2) {
      for (let i = 0; i < 100; i++) {
        let y = this.star[i].y - 3;
        if (y < 0) y += 300;
        this.star[i].y = y;
      }

      if (++this.ct > this.ct_to) {
        this.ct = 0;
        this.flash = !this.flash;
      }
      if (this.flash) {
        const mes = (this.game.score < 1000) ? "Press RESTART" : "Go to Bonus Level!";
        g.font = "bold 27px Arial, sans-serif";
        this.drawText(g, mes, { x: 200, y: 100 },
                      KLevel.CENTER, KLevel.TOP, [255, 0, 0], true);
      }
    }
  }
}
